File: frontend/src/services/currency.ts
import { api } from './api';
import { Currency, ExchangeRate } from '../types';

export const currencyService = {
  // Get all currencies
  getCurrencies: async (): Promise<Currency[]> => {
    const response = await api.get<Currency[]>('/api/currencies/');
    return response.data;
  },

  // Get exchange rate between two currencies
  getExchangeRate: async (fromCurrency: string, toCurrency: string): Promise<number> => {
    if (fromCurrency === toCurrency) return 1;

    try {
      // Use the convert endpoint instead
      const response = await api.get<any>(
        `/api/currencies/convert/1/${fromCurrency}/${toCurrency}`
      );

      return response.data.converted_amount;
    } catch (error) {
      console.error('Failed to get exchange rate:', error);
      throw new Error(`Exchange rate not found for ${fromCurrency} to ${toCurrency}`);
    }
  },

  // Convert amount between currencies
  convertAmount: async (amount: number, fromCurrency: string, toCurrency: string): Promise<number> => {
    if (fromCurrency === toCurrency) return amount;

    const rate = await currencyService.getExchangeRate(fromCurrency, toCurrency);
    return amount * rate;
  }
};

File: backend/app/routers/currency.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List
from app.database import get_db
from app.models.currency import Currency, ExchangeRate
from app.services.currency_service import CurrencyService

router = APIRouter(prefix="/api/currencies", tags=["currencies"])

@router.get("/")
async def get_currencies(db: Session = Depends(get_db)):
    """Get all available currencies"""
    currencies = db.query(Currency).filter(Currency.is_active == True).all()
    return currencies

@router.get("/{currency_code}")
async def get_currency(currency_code: str, db: Session = Depends(get_db)):
    """Get specific currency details"""
    currency = db.query(Currency).filter(Currency.code == currency_code, Currency.is_active == True).first()
    if not currency:
        raise HTTPException(status_code=404, detail="Currency not found")
    return currency

@router.get("/convert/{amount}/{from_currency}/{to_currency}")
async def convert_currency(
    amount: float, 
    from_currency: str, 
    to_currency: str,
    db: Session = Depends(get_db)
):
    """Convert amount between currencies"""
    service = CurrencyService(db)
    converted = await service.convert_amount(amount, from_currency, to_currency)
    return {
        "original_amount": amount,
        "from_currency": from_currency,
        "to_currency": to_currency,
        "converted_amount": converted
    }
